docs/wagtail-migrations.md

```diff
@@ -1,12 +1,97 @@
-# Wagtail and Django data migrations
+# Wagtail Migrations
 
-Django data migrations with Wagtail can be challenging because programmatic editing of Wagtail pages [is difficult](https://github.com/torchbox/wagtail/issues/1101), and pages have both revisions and StreamFields. This document is intended to describe ways we try to address these challenges in cfgov-refresh.
+Adding or changing fields on Wagtail page models or StreamField block classes
+will always require a new [Django schema migration](#schema-migrations);
+additionally, changing field names or types on an existing block will require a
+[Django data migration](#data-migrations).
 
-## Migrating StreamFields
 
-StreamFields do not follow a fixed structure, rather they're a freeform sequences of blocks. Making a change to a StreamField involves both creating a [Django schema migration](https://docs.djangoproject.com/en/1.11/topics/migrations/#workflow) and a custom [Django data migration](https://docs.djangoproject.com/en/1.11/topics/migrations/#data-migrations). The data migration needs to modify both the existing Wagtail pages that correspond to the changed model and all revisions of that page. It also needs to be able to manipulate the StreamField contents.
+## Table of contents
 
-To this end, there are some utility functions in cfgov-refresh that make this easier. Using these utilities, a Django data migration that modifies a StreamField would use the following format:
+1. [Reference material](#reference-material)
+1. [Schema migrations](#schema-migrations)
+1. [Data migrations](#data-migrations)
+   1. [Utility functions](#utility-functions)
+
+
+## Reference material
+
+The following links may be useful for setting context or diving deeper
+into the concepts presented throughout this page:
+
+- [Django migrations documentation](https://docs.djangoproject.com/en/1.11/topics/migrations/)
+- [Django data migrations documentation](https://docs.djangoproject.com/en/1.11/topics/migrations/#data-migrations)
+- [Wagtail Streamfield migrations documentation](https://docs.wagtail.io/en/v1.13.4/topics/streamfield.html#migrations)
+
+
+## Schema migrations
+
+Any time you are working with StreamField block or Wagtail page definitions,
+a Django schema migration will be required.
+
+To automatically generate a schema migration,
+run the following, editing it to give your migration a name
+that briefly describes the change(s) you're making:
+
+```bash
+./cfgov/manage.py makemigrations -n <description_of_changes>
+```
+
+It is a good idea to wait to do this until just before merging your new code
+into the `master` branch, to avoid potential conflicts with other developers'
+migrations that might make it into `master` before yours.
```

Comment:
I suspected this might be controversial... but in my experience, rarely is it true that you need the migration to be able to test the changes. (I don't understand _why_ that is.) But I suppose it would be safer to not say that and just tell people to look out for migration conflicts.

Reply:
You don't need a migration to test StreamField changes, but you do need one to test model field changes. Again, StreamFields are annoyingly *special*.

Reply:
> I don't understand _why_ that is.

This is likely because you are testing using `tox -e fast`, which uses [test Django settings](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/cfgov/settings/test_nomigrations.py#L6) that set [`settings.MIGRATION_MODULES`](https://docs.djangoproject.com/en/1.11/ref/settings/#migration-modules) to `None`, which does this:

> When you supply None as a value for an app, Django will consider the app as an app without migrations regardless of an existing migrations submodule. This can be used, for example, in a test settings file to skip migrations while testing (tables will still be created for the apps’ models).

In this case, the lack of migration doesn't affect the tests, and the database tables are created based on`models.py`. If you run plain `tox`, which does run migrations, and creates database tables based on those migrations, then the missing migration might cause problems with tests that access those tables.

Reply:
Oh, true! Didn't think of that either. I was thinking of when you `runserver`.

Reply:
Yeah, I was also only referring to `runserver` testing.

Reply:
Addressed in 968e783c4997ad98d51c1f8310159a1a68772b07

Reply:
Addressed **beautifully**. ❤️